### PR TITLE
Refactors url List() to use URL Client

### PR DESCRIPTION
**Pull request by mik-dass**
**What type of PR is this?**

/kind code-refactoring

**What does does this PR do / why we need it**:

Refactors url List() to use URL Client. It also adds ListCluster() to get urls from the cluster.

**Which issue(s) this PR fixes**:

Fixes part of #4298 

**PR acceptance criteria**:

- [X] Unit test 

- [ ] Integration test 

- [ ] Documentation 

- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

- Tests should pass.

### CODE REVIEW

File: pkg/odo/cli/component/status.go [hunk 1]
```diff
@@ -70,18 +70,15 @@ func (so *StatusOptions) Complete(name string, cmd *cobra.Command, args []string
 
 	// If devfile is present
 	if so.isDevfile {
-		envinfo, err := envinfo.NewEnvSpecificInfo(so.componentContext)
+		envSpecificInfo, err := envinfo.NewEnvSpecificInfo(so.componentContext)
```
Comment (dharmit) [inline pkg/odo/cli/component/status.go:73]:
```suggestion
		so.EnvSpecificInfo, err := envinfo.NewEnvSpecificInfo(so.componentContext)
```

File: pkg/url/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	routev1 "github.com/openshift/api/route/v1"
+	componentlabels "github.com/openshift/odo/pkg/component/labels"
+	"github.com/openshift/odo/pkg/localConfigProvider"
+	"github.com/openshift/odo/pkg/occlient"
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+)
+
+type kubernetesClient struct {
+	generic
```
Comment (dharmit) [inline pkg/url/kubernetes.go:16]: What is the significance of this? :thinking:

File: pkg/url/kubernetes.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	isRouteSupported bool
+	client           occlient.Client
+}
+
+// ListCluster lists both route and ingress based URLs from the cluster
+func (k kubernetesClient) ListCluster() (URLList, error) {
```
Comment (dharmit) [inline pkg/url/kubernetes.go:22]: Should the function name be something else? It lists "URLs from the cluster" but the function name seems to indicate that it lists Cluster.

File: pkg/url/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		}
+		clusterURL := getMachineReadableFormat(r)
+		clusterURLs = append(clusterURLs, clusterURL)
+	}
+
+	return getMachineReadableFormatForList(clusterURLs), nil
+}
+
+// List lists both route/ingress based URLs and local URLs with respective states
+func (k kubernetesClient) List() (URLList, error) {
```
Comment (dharmit) [inline pkg/url/kubernetes.go:56]: Can we add more one line comments in the code? At least for the logical blocks that you have created? That would help someone to read this code in the future.

**Review by dharmit — CHANGES_REQUESTED** [review]

File: pkg/url/kubernetes.go (hunk 4 above)
Comment (mik-dass) [reply pkg/url/kubernetes.go:57]: Fixed

File: pkg/url/kubernetes.go (hunk 3 above)
Comment (mik-dass) [reply pkg/url/kubernetes.go]: Renamed to `ListFromCluster()`

File: pkg/url/kubernetes.go (hunk 2 above)
Comment (mik-dass) [reply pkg/url/kubernetes.go:17]: Added comments

File: pkg/odo/cli/component/status.go (hunk 1 above)
Comment (mik-dass) [reply pkg/odo/cli/component/status.go]: Fixed

File: pkg/devfile/adapters/docker/component/utils_test.go [hunk 5]
```diff
@@ -324,15 +324,15 @@ func TestStartContainer(t *testing.T) {
 
 func TestGenerateAndGetHostConfig(t *testing.T) {
 	fakeClient := lclient.FakeNew()
-	testComponentName := "test"
+	//testComponentName := "test"
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/devfile/adapters/docker/component/utils_test.go]: Removed the test completely

File: pkg/component/component.go [hunk 6]
```diff
@@ -611,6 +610,7 @@ func ApplyConfig(client *occlient.Client, kClient *kclient.Client, componentConf
 		client.Namespace = kClient.Namespace
 	}
 
+	var localConfig localConfigProvider.LocalConfigProvider
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/component/component.go]: Fixed

File: pkg/url/kubernetes.go (hunk 3 above)
Comment (dharmit) [reply pkg/url/kubernetes.go]: Thanks. That makes much more sense than the previous name. :+1:

**Review by dharmit — APPROVED** [review]
/approve
/lgtm

@mik-dass thanks for addressing the requests. :bowing_man: :+1:

**Comment by mik-dass** [comment]
```
Should be able to verify the nodejs-12-rhel7 image [It]
    /go/src/github.com/openshift/odo/tests/e2escenarios/e2e_images_test.go:137
    Failed after 90 retries. Content in  doesn't include 'Hello'.
```
/hold

**Comment by mik-dass** [comment]
```
STEP 1: FROM registry.svc.ci.openshift.org/openshift/release:golang-1.13
error: build error: error creating build container: Error initializing source docker://registry.svc.ci.openshift.org/openshift/release:golang-1.13: Get "https://registry.svc.ci.openshift.org/openshift/token?account=serviceaccount&scope=repository%3Aopenshift%2Frelease%3Apull": dial tcp 35.196.103.194:443: connect: no route to host
```
/retest

**Comment by mik-dass** [comment]
```
error: build error: error creating build container: Error initializing source docker://registry.svc.ci.openshift.org/openshift/release:golang-1.13: Get "https://registry.svc.ci.openshift.org/openshift/token?account=serviceaccount&scope=repository%3Aopenshift%2Frelease%3Apull": dial tcp 35.196.103.194:443: connect: no route to host
```
/retest

**Comment by mik-dass** [comment]
```
th args [oc exec eqnbje-97cd44cfd-4ncgk -c runtime --namespace anwidomgtr -- ps -ef]
[oc] Signal 23 (URG) caught by ps (3.3.15).
[oc] ps:ps/display.c:66: please report this bug
[oc] command terminated with exit code 1
Deleting project: anwidomgtr
```
/retest

**Comment by dharmit** [comment]
@mik-dass can you address the conflicts? I'd try to get this in before the end of this Sprint.

**Comment by mik-dass** [comment]
> @mik-dass can you address the conflicts? I'd try to get this in before the end of this Sprint. Also, please remove the hold if things are good.

Fixed

**Review by dharmit — APPROVED** [review]
/lgtm

**Pull request merged**